icechunk/src/session.rs

```diff
@@ -0,0 +1,1055 @@
+use std::{
+    collections::HashSet,
+    future::{ready, Future},
+    iter,
+    mem::take,
+    pin::Pin,
+    sync::Arc,
+};
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use chrono::Utc;
+use futures::{FutureExt, Stream, StreamExt, TryStreamExt};
+use thiserror::Error;
+use tokio::sync::RwLock;
+
+use crate::{
+    change_set::ChangeSet,
+    conflicts::{ConflictResolution, ConflictSolver},
+    format::{
+        manifest::{
+            ChunkInfo, ChunkRef, Manifest, ManifestExtents, ManifestRef, VirtualChunkRef,
+        },
+        snapshot::{
+            ManifestFileInfo, NodeData, NodeSnapshot, NodeType, Snapshot,
+            SnapshotProperties, UserAttributesSnapshot,
+        },
+        transaction_log::TransactionLog,
+        ByteRange, ChunkIndices, IcechunkFormatError, ManifestId, NodeId, Path,
+        SnapshotId,
+    },
+    metadata::UserAttributes,
+    refs::{fetch_branch_tip, update_branch, RefError},
+    repository::{ChunkPayload, RepositoryError, RepositoryResult, ZarrArrayMetadata},
+    storage::virtual_ref::{construct_valid_byte_range, VirtualChunkResolver},
+    store::Store,
+    zarr::{ObjectId, StoreOptions},
+    RepositoryConfig, Storage,
+};
+
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum SessionError {
+    #[error("Read only sessions cannot modify the repository")]
+    ReadOnlySession,
+    #[error("Repository error: {0}")]
+    RepositoryError(#[from] RepositoryError),
+}
+
+pub type SessionResult<T> = Result<T, SessionError>;
+
+pub struct Session {
+    config: RepositoryConfig,
+    storage: Arc<dyn Storage + Send + Sync>,
+    virtual_resolver: Arc<dyn VirtualChunkResolver + Send + Sync>,
+    branch_name: Option<String>,
+    snapshot_id: SnapshotId,
+    change_set: ChangeSet,
+}
+
+impl Session {
+    pub fn create_readable_session(
+        config: RepositoryConfig,
+        storage: Arc<dyn Storage + Send + Sync>,
+        virtual_resolver: Arc<dyn VirtualChunkResolver + Send + Sync>,
+        snapshot_id: SnapshotId,
+    ) -> Self {
+        Self {
+            config,
+            storage,
+            virtual_resolver,
+            branch_name: None,
+            snapshot_id,
+            change_set: ChangeSet::default(),
+        }
+    }
+
+    pub fn create_writable_session(
+        config: RepositoryConfig,
+        storage: Arc<dyn Storage + Send + Sync>,
+        virtual_resolver: Arc<dyn VirtualChunkResolver + Send + Sync>,
+        branch_name: String,
+        snapshot_id: SnapshotId,
+    ) -> Self {
+        Self {
+            config,
+            storage,
+            virtual_resolver,
+            branch_name: Some(branch_name),
+            snapshot_id,
+            change_set: ChangeSet::default(),
+        }
+    }
+
+    pub fn read_only(&self) -> bool {
+        self.branch_name.is_none()
+    }
+
+    pub fn store(&mut self) -> Store {
+        let read_only = self.read_only();
+        let mutexed = Arc::new(RwLock::new(self));
```

Comment:
this looks quite suspicious to me, but haven't thought about it enough. This when calling `store`, you give out the only possible mutable reference to the Session, so you no longer can make changes to the session until the `Store` is freed. Will have to wrap my head around it, not sure how commit would work for example.

Reply:
I really don't like this. I don't know how else to manage though. Need to think through the hierarchy

Reply:
OK the only mutable part is the `ChangeSet` I think, so we can in theory just `Arc<Mutex` the changeset and then the session is no longer locked. I think